src/Utilities.UnitTests/Options/SymbolNamesWithValueOptionTests.cs

```diff
@@ -0,0 +1,361 @@
+        [InlineData("MyCompany.MyProduct*", "MyProduct")]
+        [InlineData("MyCompany.MyProduct.MyFeature*", "MyFeature")]
+        [InlineData("MyCompany.MyProduct.MyFeature.MyOuterClass*", "MyOuterClass")]
+        [InlineData("MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass*", "MyInnerClass")]
+        [InlineData("MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyField*", "MyField")]
+        [InlineData("MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyProperty*", "MyProperty")]
+        [InlineData("MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyEvent*", "MyEvent")]
+        [InlineData("MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyMethod(*", "MyMethod")]
+        // Fully qualified name with prefix
+        [InlineData("N:MyCompany*", "MyCompany")]
+        [InlineData("N:MyCompany.MyProduct*", "MyProduct")]
+        [InlineData("N:MyCompany.MyProduct.MyFeature*", "MyFeature")]
+        [InlineData("T:MyCompany.MyProduct.MyFeature.MyOuterClass*", "MyOuterClass")]
+        [InlineData("T:MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass*", "MyInnerClass")]
+        [InlineData("F:MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyField*", "MyField")]
+        [InlineData("P:MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyProperty*", "MyProperty")]
+        [InlineData("E:MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyEvent*", "MyEvent")]
+        [InlineData("M:MyCompany.MyProduct.MyFeature.MyOuterClass.MyInnerClass.MyMethod(*", "MyMethod")]
+        // Partial match
+        [InlineData("MyCompany.MyProduct*", "MyFeature")]
+        [InlineData("N:MyCompany.MyProduct*", "MyFeature")]
+        [InlineData("MyFeature*", "MyFeature")]
+        [InlineData("MyInnerClass*", "MyInnerClass")]
+        [InlineData("T:MyInnerClass*", "MyInnerClass")]
+        [InlineData("MyOuterClass*", "MyInnerClass")]
+        [InlineData("T:MyOuterClass*", "MyInnerClass")]
+        public void WildcardMatch(string patternName, string symbolName)
+        {
+            // Arrange
+            var compilation = GetCompilation(@"
+public namespace MyCompany.MyProduct.MyFeature
+{
+    public class MyOuterClass
+    {
+        public class MyInnerClass
+        {
+            public int MyField;
+            public int MyProperty { get; set; }
+            public event EventHandler<EventArgs> MyEvent;
+            public void MyMethod() {}
+        }
+    }
+}");
+            var symbolNames = ImmutableArray.Create(patternName);
+
+            var symbol = compilation.GetSymbolsWithName(symbolName, SymbolFilter.All).Single();
+
+            var options = SymbolNamesWithValueOption<Unit>.Create(symbolNames, compilation, null, null);
+
+            // Act
+            var isFound = options.Contains(symbol);
+
+            // Assert
+            Assert.True(isFound);
+            Assert.True(options._wildcardMatchResult.ContainsKey(symbol));
+        }
```

Comment:
Thanks for writing these exhaustive unit tests!

Reply:
You are welcome. The code I have added is too complex so adding tests only through the actual rules wasn't giving me enough confidence.